VSS.md

```diff
@@ -0,0 +1,484 @@
+## RFC: 8
+
+## Status: Proposed
+
+
+# Valkey Search RFC
+
+## Abstract
+
+This proposed Valkey Search module will introduce search algorithms to provide numeric, tag and vector search capabilities on both Valkey Cluster and Valkey Standalone. With search on Valkey, users can take advantage of low latency in-memory search in their familiar Valkey environment.
+
+## Motivation
+
+With the rising popularity of generative AI, developers have turned to in-memory data stores such as Valkey or Redis for ultra-low latency (p99 single digit ms) vector search for use cases like Retrieval Augmented Generation (RAG), recommendation systems, semantic search, semantic caching, and more. For developers already using Valkey or Redis, co-locating their vectors and search functionality alongside their underlying data will accelerate and simplify adoption. And for other “greenfield” users, the ultra-low latency of Valkey search may attract them to replace their existing  database for high performance use cases to achieve in-memory speeds.
+
+## Design considerations
+
+Here we list Critical User Journeys (CUJs) which we believe Valkey users will desire as part of a Search Module.
+
+**Search Operations CUJ** \- as a user, I want to perform search operations using commonly available clients.
+
+- Requirement: support index insertion/mutation/deletion/query operations  
+- Requirement: maintain compatibility with RediSearch VSS APIs, Memorystore APIs, and MemoryDB APIs as much as possible
```

Comment:
RedisSearch API is weird. Pinecone/Milvus API is more natural
```
# create
pinecone.create_index(name=index_name, dimension=dimension)
# insert
index = pinecone.Index(index_name)
index.upsert(vectors)
# search
results = index.query(queries=[query_vector], top_k=5)
```

I believe developers prefer the latter.

Reply:
So to clarify the difference:
1. In the current implementation, you insert keys with other commands and then create an index ontop of a prefix structure.
2. In the pinecone world, you create an index and then insert objects into the index explicitly.

?

Reply:
I agree that the API is not intuitive and has room for improvement. However, we've opted for RediSearch compatibility to facilitate easier adoption. In the long run, though, we should explore designing a more user-friendly API.